profile/filter_test.go

```diff
@@ -0,0 +1,252 @@
+		{ID: 10, Mapping: mappings[0], Address: 0x10000, Line: []Line{{Function: functions[9], Line: 1}}},
+		{ID: 11, Mapping: mappings[1], Address: 0x11000, Line: []Line{{Function: functions[10], Line: 1}}},
+	}
+	profile := &Profile{
+		TimeNanos:     10000,
+		PeriodType:    &ValueType{Type: "cpu", Unit: "milliseconds"},
+		Period:        1,
+		DurationNanos: 10e9,
+		SampleType: []*ValueType{{Type: "samples", Unit: "count"}},
+		Mapping:  mappings,
+		Function: functions,
+		Location: locations,
+		Sample: []*Sample{
+			{Value: []int64{1}, Location: []*Location{locations[0], locations[1], locations[2], locations[3]}},
+			{Value: []int64{2}, Location: []*Location{locations[4], locations[5], locations[1], locations[6]}},
+			{Value: []int64{3}, Location: []*Location{locations[7], locations[8]}},
+			{Value: []int64{4}, Location: []*Location{locations[9], locations[4], locations[10], locations[7]}},
+		},
+	}
+
+	for _, testCase := range []struct{
+		name string
+		focus, ignore, hide, show *regexp.Regexp
+		// todo: fm, im, hm, hnm           bool
+		wantPaths [][]int // expected paths. wantPaths[sampleIndex][frameIndex] = locationIndex
+		wantValues []int64 // expected values. wantValues[sampleIndex] = sampleValue
+	}{
+		{
+			name: "empty filters keep all frames",
+			// todo: fm: true, // nil focus matches every sample
+			wantPaths: [][]int{
+				{0, 1, 2, 3},
+				{4, 5, 1, 6},
+				{7, 8},
+				{9, 4, 10, 7},
+			},
+			wantValues: []int64{1, 2, 3, 4},
+		},
+		{
+			name: "hides a function if both show and hide match it",
+			show: regexp.MustCompile("fun1"),
+			hide: regexp.MustCompile("fun10"),
+			wantPaths: [][]int{
+				{1},
+				{1},
+			},
+			wantValues: []int64{1, 2},
+		},
+	} {
+		t.Run(testCase.name, func(t *testing.T) {
+			gotProfile := profile.Copy()
+			gotProfile.FilterSamplesByName(testCase.focus, testCase.ignore, testCase.hide, testCase.show)
+			gotStr := gotProfile.String()
+
+			wantSamples, wantLocations := makeFilteredSamples(testCase.wantPaths, testCase.wantValues, locations)
+			wantProfile := profile.Copy()
+			wantProfile.Sample = wantSamples
+			wantProfile.Location = wantLocations
+			wantStr := wantProfile.String()
+
+			if wantStr != gotStr {
+				diff, err := proftest.Diff([]byte(wantStr), []byte(gotStr))
+				if err != nil {
+					t.Fatalf("failed to get diff: %v", err)
+				}
+				t.Errorf("filtered profile got diff(want->got): \n%s", diff)
+			}
+		})
+	}
+}
+
+// makeSamples constructs profile samples from basic inputs.
+// paths: location indices, indexed by sample and frame.
+// 	i.e. paths[sampleIndex][frameIndex] = locationIndex
+// values: values for each sample. i.e. values[sampleIndex] = value
+// locations: locations to copy for new samples. unused locations are not
+// 	removed but their lines are set to nil.
+func makeFilteredSamples(paths [][]int, values []int64, locations []*Location) ([]*Sample, []*Location) {
```

Comment:
This helper feels over-complex, I am still reading through its code. I would consider a clearer (even if a bit less accurate) validation of the expected result. For example, walk over profile and compute a simple flat report (funcName, flat, cum) similar to how printTopProto in internal/report/report.go does. And express the expected result as a list of rows of that type, like [("fun1", 200, 500), "fun2, 300, 600)] or like that. With that and a bit smaller number of functions I think the test would end up being more readable and easier extensible over time, if needed.

Reply:
That doesn't seem worth it, imo. It sounds like its just as much work to create and maintain a less accurate test.